File: backend/src/projects/projects.service.ts
import ProjectsModel from "./projects.model";
import { logger } from "../config/logger";

export const createProjectService = async (
  userId: string,
  projectData: { projectName: string; summary: string; primaryStack: string; status: string; deadline: string; teamSize: string; projectNotes: string; focusTags?: string[]; progress?: number }
) => {
  try {
    logger.info("Creating project for user => " + userId);
    const project = await ProjectsModel.create({
      userId,
      ...projectData,
    });
    logger.info("Project created successfully => " + project._id);
    return project;
  } catch (error) {
    logger.error("Project creation failed => " + error);
    throw Error("Project creation failed in service");
  }
};

export const getAllProjectsService = async (userId: string) => {
  try {
    logger.info("Fetching all projects for user => " + userId);
    // Filter deletedAt: null to exclude soft-deleted projects
    const projects = await ProjectsModel.find({ userId, deletedAt: null }).sort({ createdAt: -1 });
    logger.info("Projects fetched successfully, total => " + projects.length);
    return projects;
  } catch (error) {
    logger.error("Failed to fetch projects => " + error);
    throw Error("Failed to fetch projects in service");
  }
};

export const getProjectByIdService = async (projectId: string, userId: string) => {
  try {
    logger.info("Fetching project => " + projectId);
    // Ensure project belongs to user and is not deleted
    const project = await ProjectsModel.findOne({ _id: projectId, userId, deletedAt: null });
    if (!project) {
      logger.error("Project not found => " + projectId);
      return null;
    }
    logger.info("Project fetched successfully");
    return project;
  } catch (error) {
    logger.error("Failed to fetch project => " + error);
    throw Error("Failed to fetch project in service");
  }
};

export const updateProjectService = async (
  projectId: string,
  userId: string,
  updateData: { projectName?: string; summary?: string; primaryStack?: string; status?: string; deadline?: string; teamSize?: string; projectNotes?: string; focusTags?: string[]; progress?: number }
) => {
  try {
    logger.info("Updating project => " + projectId + " for user => " + userId);
    // Using findOneAndUpdate with new: true to return updated document
    const project = await ProjectsModel.findOneAndUpdate(
      { _id: projectId, userId, deletedAt: null },
      updateData,
      { new: true }
    );
    if (!project) {
      logger.error("Project not found for update => " + projectId);
      return null;
    }
    logger.info("Project updated successfully => " + projectId);
    return project;
  } catch (error) {
    logger.error("Failed to update project => " + error);
    throw Error("Failed to update project in service");
  }
};

export const deleteProjectService = async (projectId: string, userId: string) => {
  try {
    logger.info("Deleting project => " + projectId + " for user => " + userId);
    // Soft delete: setting deletedAt instead of removing from DB
    const project = await ProjectsModel.findOneAndUpdate(
      { _id: projectId, userId, deletedAt: null },
      { deletedAt: new Date() },
      { new: true }
    );
    if (!project) {
      logger.error("Project not found for deletion => " + projectId);
      return null;
    }
    logger.info("Project deleted successfully => " + projectId);
    return project;
  } catch (error) {
    logger.error("Failed to delete project => " + error);
    throw Error("Failed to delete project in service");
  }
};

File: frontend/dev_vault/lib/data/services/projects_service.dart
import 'package:dio/dio.dart';
import '../constant_urls.dart';
import '../models/projects_model.dart';
import 'local_storage_service.dart';

class ProjectsService {
  static final Dio _dio = Dio();

  static Map<String, dynamic> _getAuthHeaders() {
    final token = LocalStorageService.getToken();
    return {'Authorization': token != null ? 'Bearer $token' : ''};
  }

  /// Get all projects
  static Future<List<ProjectsModel>> getAllProjects() async {
    try {
      print('📡 Fetching all projects from: $getAllProjectsUrl');

      final response = await _dio.get(
        getAllProjectsUrl,
        options: Options(headers: _getAuthHeaders()),
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = response.data;

        if (data['success'] == true && data['data'] != null) {
          final projectsList = (data['data'] as List)
              .map((project) => ProjectsModel.fromJson(project))
              .toList();
          print('✅ Projects fetched: ${projectsList.length} total');
          return projectsList;
        } else {
          throw Exception(data['message'] ?? 'Failed to fetch projects');
        }
      } else {
        throw Exception('Failed with status ${response.statusCode}');
      }
    } on DioException catch (e) {
      print('❌ Error: ${e.message}');
      throw Exception(e.response?.data['message'] ?? e.message);
    } catch (e) {
      throw Exception(e.toString());
    }
  }

  /// Get project by ID
  static Future<ProjectsModel> getProjectById(String id) async {
    try {
      print('📡 Fetching project: $id');

      final response = await _dio.get(
        '$getProjectByIdUrl$id',
        options: Options(headers: _getAuthHeaders()),
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = response.data;
        if (data['success'] == true && data['data'] != null) {
          final project = ProjectsModel.fromJson(data['data']);
          print('✅ Project fetched: ${project.projectName}');
          return project;
        } else {
          throw Exception(data['message'] ?? 'Failed to fetch project');
        }
      } else {
        throw Exception('Failed with status ${response.statusCode}');
      }
    } on DioException catch (e) {
      throw Exception(e.response?.data['message'] ?? e.message);
    } catch (e) {
      throw Exception(e.toString());
    }
  }

  /// Create project
  static Future<ProjectsModel> createProject({
    required String projectName,
    required String summary,
    required String primaryStack,
    required String status,
    required String deadline,
    required String teamSize,
    required String projectNotes,
    List<String> focusTags = const [],
    int progress = 0,
  }) async {
    try {
      print('🚀 Creating project: $projectName');

      final response = await _dio.post(
        createProjectUrl,
        data: {
          'projectName': projectName,
          'summary': summary,
          'primaryStack': primaryStack,
          'status': status,
          'deadline': deadline,
          'teamSize': teamSize,
          'projectNotes': projectNotes,
          'focusTags': focusTags,
          'progress': progress,
        },
        options: Options(headers: _getAuthHeaders()),
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = response.data;
        if (data['success'] == true && data['data'] != null) {
          final project = ProjectsModel.fromJson(data['data']);
          print('✅ Project created: ${project.projectName}');
          return project;
        } else {
          throw Exception(data['message'] ?? 'Failed to create project');
        }
      } else {
        throw Exception('Failed with status ${response.statusCode}');
      }
    } on DioException catch (e) {
      throw Exception(e.response?.data['message'] ?? e.message);
    } catch (e) {
      throw Exception(e.toString());
    }
  }

  /// Update project
  static Future<ProjectsModel> updateProject({
    required String id,
    required String projectName,
    required String summary,
    required String primaryStack,
    required String status,
    required String deadline,
    required String teamSize,
    required String projectNotes,
    List<String> focusTags = const [],
    int progress = 0,
  }) async {
    try {
      print('✏️ Updating project: $id');

      final response = await _dio.put(
        '$updateProjectUrl$id',
        data: {
          'projectName': projectName,
          'summary': summary,
          'primaryStack': primaryStack,
          'status': status,
          'deadline': deadline,
          'teamSize': teamSize,
          'projectNotes': projectNotes,
          'focusTags': focusTags,
          'progress': progress,
        },
        options: Options(headers: _getAuthHeaders()),
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = response.data;
        if (data['success'] == true && data['data'] != null) {
          final project = ProjectsModel.fromJson(data['data']);
          print('✅ Project updated: ${project.projectName}');
          return project;
        } else {
          throw Exception(data['message'] ?? 'Failed to update project');
        }
      } else {
        throw Exception('Failed with status ${response.statusCode}');
      }
    } on DioException catch (e) {
      throw Exception(e.response?.data['message'] ?? e.message);
    } catch (e) {
      throw Exception(e.toString());
    }
  }

  /// Delete project
  static Future<void> deleteProject(String id) async {
    try {
      print('🗑️ Deleting project: $id');

      final response = await _dio.delete(
        '$deleteProjectUrl$id',
        options: Options(headers: _getAuthHeaders()),
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = response.data;
        if (data['success'] == true) {
          print('✅ Project deleted');
        } else {
          throw Exception(data['message'] ?? 'Failed to delete project');
        }
      } else {
        throw Exception('Failed with status ${response.statusCode}');
      }
    } on DioException catch (e) {
      throw Exception(e.response?.data['message'] ?? e.message);
    } catch (e) {
      throw Exception(e.toString());
    }
  }
}
